File: plugins/ads/source/Android/src/com/giderosmobile/android/plugins/ads/frameworks/AdsUnity.java
package com.giderosmobile.android.plugins.ads.frameworks;

import java.lang.ref.WeakReference;
import android.app.Activity;
import android.util.SparseArray;
import android.view.KeyEvent;

import com.giderosmobile.android.plugins.ads.*;
import com.unity3d.ads.*;


public class AdsUnity implements AdsInterface, IUnityAdsListener {

	private WeakReference<Activity> sActivity;
	private AdsManager mngr;
	private static AdsUnity me;
	private boolean v4vcMap=true;
	
	private String mapType(String type)
	{
		if (type.equals("rewardedVideo"))
			v4vcMap=false;
		if (v4vcMap&&type.equals("v4vc"))
			return "rewardedVideo";
		return type;
	}
	
	private String unmapType(String type)
	{
		if (v4vcMap&&type.equals("rewardedVideo"))
			return "v4vc";
		return type;		
	}

	public void onCreate(WeakReference<Activity> activity)
	{
		me = this;
		sActivity = activity;
		mngr = new AdsManager();
	}

	//on destroy event
	public void onDestroy(){}

	public void onStart(){}

	public void onStop(){}

	public void onPause()
	{
	}

	public void onResume()
	{
	}

	public boolean onKeyUp(int keyCode, KeyEvent event) {
		return false;
	}

	public void setKey(final Object parameters){
		SparseArray<String> param = (SparseArray<String>)parameters;
		UnityAds.initialize(sActivity.get(), param.get(0), this);
	}


	//load an Ad
	public void loadAd(final Object parameters)
	{
		final SparseArray<String> param = (SparseArray<String>)parameters;
		final String type = param.get(0);
			mngr.set(UnityAds.class, type, new AdsStateChangeListener(){
				@Override
				public void onShow() {
					UnityAds.show(sActivity.get(),mapType(type));
				}
				@Override
				public void onDestroy() {}
				@Override
				public void onHide() {}
				@Override
                public void onRefresh() {}
			});
			mngr.load(type);
	}

	public void showAd(final Object parameters)
	{
		SparseArray<String> param = (SparseArray<String>)parameters;
		String type = param.get(0);
		if(mngr.get(type) == null)
		{
			loadAd(parameters);
		}
		mngr.show(type);
	}

	//remove ad
	public void hideAd(String type)
	{
		mngr.hide(type);
	}

	public int getWidth(){
		return 0;
	}

	public int getHeight(){
		return 0;
	}

	@Override
	public void enableTesting() {

	}

	@Override
	public void onUnityAdsReady(String s) {
		Ads.adReceived(this,unmapType(s));
	}

	@Override
	public void onUnityAdsStart(String s) {
		Ads.adDisplayed(this,unmapType(s));
	}

	@Override
	public void onUnityAdsFinish(String s, UnityAds.FinishState finishState) {
		if (finishState== UnityAds.FinishState.COMPLETED)
			Ads.adActionEnd(this,unmapType(s));
		else if (finishState== UnityAds.FinishState.SKIPPED)
			Ads.adDismissed(this,unmapType(s));
		else if (finishState== UnityAds.FinishState.ERROR)
			Ads.adFailed(this,unmapType(s),"");
	}

	@Override
	public void onUnityAdsError(UnityAds.UnityAdsError unityAdsError, String s) {
	}
}

File: plugins/ads/source/iOS/Plugins/Ads/frameworks/AdsUnity.m
//
//  AdsAdcolony.m
//  Ads
//
//  Created by Arturs Sosins on 6/25/13.
//  Copyright (c) 2013 Gideros Mobile. All rights reserved.
//
#include "gideros.h"
#import "AdsUnity.h"
#import "AdsClass.h"


@implementation AdsUnity

-(id)init{
    self.mngr = [[AdsManager alloc] init];
    self.v4vcMap=YES;
    return self;
}

-(void)destroy{
    [self.mngr destroy];
    [self.mngr release];
    self.mngr = nil;
}

-(void)setKey:(NSMutableArray*)parameters{
    [UnityAds initialize:[parameters objectAtIndex:0] delegate: self];
}

- (NSString *) mapType:(NSString *)type
{
		if ([type isEqual:@"rewardedVideo"])
			_v4vcMap=NO;
		if (_v4vcMap&&[type isEqual:@"v4vc"])
			return @"rewardedVideo";
		return type;
}
	
- (NSString *) unmapType:(NSString *)type
{
		if (_v4vcMap&&[type isEqual:@"rewardedVideo"])
			return @"v4vc";
		return type;		
}

#pragma mark - UnityAds Delegate

- (void)unityAdsReady:(NSString *)placementId {
    [AdsClass adReceived:[self class] forType:[self unmapType:placementId]];
}

- (void)unityAdsDidStart:(NSString *)placementId {
    NSLog(@"An ad is about to be played!");
    [AdsClass adDisplayed:[self class] forType:[self unmapType:placementId]];
}

- (void)unityAdsDidError:(UnityAdsError)error withMessage:(NSString *)message {
    NSLog(@"Unity Ads Error occured:%@",message);
}

- (void)unityAdsDidFinish:(NSString *)placementId withFinishState:(UnityAdsFinishState)state {
    if(state==kUnityAdsFinishStateCompleted){
        [AdsClass adActionEnd:[self class] forType:[self unmapType:placementId]];
    } else if(state==kUnityAdsFinishStateSkipped) {
        [AdsClass adDismissed:[self class] forType:[self unmapType:placementId]];
    } else if(state==kUnityAdsFinishStateError) {
        [AdsClass adFailed:[self class] with:@"" forType:[self unmapType:placementId]];
    }
}


-(void)loadAd:(NSMutableArray*)parameters{
    NSString *type = [parameters objectAtIndex:0];

        AdsStateChangeListener *listener = [[AdsStateChangeListener alloc] init];
            [listener setShow:^(){
                [UnityAds show:[AdsClass getRootViewController] placementId:[self mapType:type]];
            }];
            [listener setDestroy:^(){}];
            [listener setHide:^(){}];
            [self.mngr set:type forType:type withListener:listener];
            [self.mngr load:type];
 }

-(void)showAd:(NSMutableArray*)parameters{
    NSString *type = [parameters objectAtIndex:0];
    if([self.mngr get:type] == nil)
        [self loadAd:parameters];
    [self.mngr show:type];
}

-(void)hideAd:(NSString*)type{
    [self.mngr hide:type];
}

-(void)enableTesting{
}

-(UIView*)getView{
    return nil;
}
@end
